agents/hwp_gripper/gripper_agent.py

```diff
@@ -0,0 +1,575 @@
+#!/usr/bin/env python3
+
+import argparse
+import ctypes
+import multiprocessing
+import os
+import signal
+import subprocess
+import sys
+import time
+
+import numpy as np
+
+this_dir = os.path.dirname(__file__)
+sys.path.append(
+    os.path.join(this_dir, 'src'))
+sys.path.append(
+    os.path.join(this_dir, 'pru'))
+
+import gripper_client as gclient
+import GripperBuilder as gb
+import GripperCollector as gc
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import TimeoutLock
+
+# from socs.agents.hwp_supervisor.agent import get_op_data
+
+
+class GripperAgent:
+    def __init__(self, agent, mcu_ip, pru_port, control_port,
+                 return_port, supervisor_id=None, no_data_timeout=45 * 60):
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+
+        self._initialized = False
+        self.mcu_ip = mcu_ip
+        self.pru_port = pru_port
+        self.control_port = control_port
+        self.return_port = return_port
+
+        self.shutdown_mode = False
+        self.supervisor_id = supervisor_id
+        self.no_data_timeout = no_data_timeout
+
+        self.collector = None
+        self.builder = None
+        self.client = None
+
+        self.last_encoder = 0
+        self.last_limit = 0
+        self.last_limit_time = 0
+        self.is_forced = False
+
+        # self.encoder_names = ['Actuator 1 A', 'Actuator 1 B', 'Actuator 2 A',
+        #                      'Actuator 2 B', 'Actuator 3 A', 'Actuator 3 B']
+
+        self.encoder_pru = [0, 1, 2, 3, 4, 5]
+        self.encoder_edges = multiprocessing.Array(ctypes.c_int, (0, 0, 0, 0, 0, 0))
+        self.encoder_edges_record = multiprocessing.Array(ctypes.c_int, (0, 0, 0, 0, 0, 0))
+        self.encoder_direction = multiprocessing.Array(ctypes.c_int, (1, 1, 1, 1, 1, 1))
+
+        self.limit_names = ['Actuator 1 Cold', 'Actuator 1 Warm', 'Actuator 2 Cold',
+                            'Actuator 2 Warm', 'Actuator 3 Cold', 'Actuator 3 Warm']
+        self.limit_pru = [8, 9, 10, 11, 12, 13]
+        self.limit_state = [0, 0, 0, 0, 0, 0]
+
+        self.mode = multiprocessing.Value(ctypes.c_bool, False)
+        self.force = multiprocessing.Value(ctypes.c_bool, False)
+
+        agg_params = {'frame_length': 60}
+        self.agent.register_feed('hwpgripper', record=True, agg_params=agg_params)
+        self.agent.register_feed('gripper_action', record=True)
+
+    @ocs_agent.param('auto_acquire', default=True, type=bool)
+    def init_processes(self, session, params):
+        if self._initialized:
+            self.log.info('Connection already initialized. Returning...')
+            return True, 'Connection already initialized'
+
+        self.client = gclient.GripperClient(self.mcu_ip, self.control_port, self.return_port)
+        self.collector = gc.GripperCollector(self.pru_port)
+
+        self.agent.start('grip_collect_pru')
+        self.agent.start('grip_build_pru')
```

Comment:
The agent requires that the server/pru processes are running already on the Beaglebone already, after which it will connect to them. If the processes are already running on the Beaglebone I have never run into connection issues.